src/Storages/StorageS3Distributed.cpp

```diff
@@ -0,0 +1,373 @@
+#include <Processors/Pipe.h>
+#include <Processors/Sources/SourceFromInputStream.h>
+#include <Parsers/queryToString.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Storages/IStorage.h>
+#include <Storages/SelectQueryInfo.h>
+#include <Storages/StorageS3.h>
+#include <common/logger_useful.h>
+
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+
+#include <ios>
+#include <memory>
+#include <string>
+#include <thread>
+#include <cassert>
+
+namespace DB
+{
+
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+struct StorageS3SourceBuilder
+{
+    bool need_path;
+    bool need_file;
+    String format;
+    String name;
+    Block sample_block;
+    const Context & context;
+    const ColumnsDescription & columns;
+    UInt64 max_block_size;
+    String compression_method;
+};
+
+class StorageS3SequentialSource : public SourceWithProgress
+{
+public:
+
+    static Block getHeader(Block sample_block, bool with_path_column, bool with_file_column)
+    {
+        if (with_path_column)
+            sample_block.insert({DataTypeString().createColumn(), std::make_shared<DataTypeString>(), "_path"});
+        if (with_file_column)
+            sample_block.insert({DataTypeString().createColumn(), std::make_shared<DataTypeString>(), "_file"});
+
+        return sample_block;
+    }
+
+    StorageS3SequentialSource(
+        String initial_query_id_,
+        Cluster::Address initiator,
+        const ClientAuthentificationBuilder & client_auth_builder_,
+        const StorageS3SourceBuilder & s3_builder_)
+        : SourceWithProgress(getHeader(s3_builder_.sample_block, s3_builder_.need_path, s3_builder_.need_file))
+        , initial_query_id(initial_query_id_)
+        , s3_source_builder(s3_builder_)
+        , cli_builder(client_auth_builder_)
+    {
+        connections = std::make_shared<ConnectionPool>(
+            /*max_connections*/3,
+            /*host*/initiator.host_name,
+            /*port*/initiator.port,
+            /*default_database=*/s3_builder_.context.getGlobalContext().getCurrentDatabase(),
+            /*user=*/s3_builder_.context.getClientInfo().initial_user,
+            /*password=*/initiator.password,
+            /*cluster=*/initiator.cluster,
+            /*cluster_secret=*/initiator.cluster_secret
+        );
+
+        createOrUpdateInnerSource();
+    }
+
+    String getName() const override
+    {
+        return "StorageS3SequentialSource"; 
+    }
+
+    Chunk generate() override
+    {
+        if (!inner)
+            return {};
+
+        auto chunk = inner->generate();
+        if (!chunk) 
+        {
+            if (!createOrUpdateInnerSource())
+                return {};
+            else
+                chunk = inner->generate();
+        }
+        return chunk;
+    }
+
+private:
+
+    String askAboutNextKey()
+    {
+        try
+        {
+            auto connection = connections->get(timeouts);
```

Comment:
So, here we get a new connection for every task ...
I think it would be more logical to just reuse an existing one

Reply:
Deleted this code. Now we ask read task in the same connection